src/service/gateways/bitstamp.ts

```diff
@@ -0,0 +1,360 @@
+            this.onOrderBookMessage(new Models.Timestamped(b, moment.utc())))
+                .done();
+            
+        restClient.getFromEndpoint<MarketTransaction[]>("transactions/").then(ts => {
+            ts.forEach(t => {
+                var side = t.type === 0 ? Models.Side.Bid : Models.Side.Ask; // is this really the make side?
+                var trd = new Models.GatewayMarketTrade(parseFloat(t.price), parseFloat(t.amount), moment(t.date, "X"), true, side);
+                this.MarketTrade.trigger(trd);
+            });
+        }).done();
+    }
+}
+
+class BitstampAuthenticatedClient {
+    private static UserAgent = 'Mozilla/4.0 (compatible; Bitstamp node.js client)';
+    
+    constructor(
+        private baseUrl: string, 
+        private _customerId: string, 
+        private _apiKey: string, 
+        private _secret: string) { }
+        
+    private _lastTimeMs : number = 0;
+    private getNonce = () => {
+        var t = new Date().getTime();
+        if (t === this._lastTimeMs) {
+            this._lastTimeMs++;
+        }
+        else {
+            this._lastTimeMs = t*100;
+        }
+        
+        return this._lastTimeMs;
+    };
+    
+    private addAuthentication = (req: {}) => {
+        var nonce = this.getNonce();
+        var message = nonce + this._customerId + this._apiKey;
+        var signer = crypto.createHmac('sha256', new Buffer(this._secret, 'utf8'));
+        var signature = signer.update(message).digest('hex').toUpperCase();
+        
+        return _.extend({
+            key: this._apiKey,
+            signature: signature,
+            nonce: nonce}, req);
+    }; 
+    
+    // the URLs must end with slashes in their API
+    private makeUrl = (endpoint: string) => this.baseUrl + "/" + endpoint;
+    
+    public postToEndpoint = <TRequest, TResponse>(endpoint: string, req : TRequest) : Q.Promise<TResponse> => {
+        return this.requestFromEndpoint<TResponse>({
+            url: this.makeUrl(endpoint),
+            form: this.addAuthentication(req),
+            method: 'POST',
+            timeout: 5000,
+            headers: { 'User-Agent': BitstampAuthenticatedClient.UserAgent }
+        });
+    }; 
+    
+    public getFromEndpoint = <TResponse>(endpoint: string) => {
+        return this.requestFromEndpoint<TResponse>({
+            url: this.makeUrl(endpoint),
+            method: 'GET',
+            timeout: 5000,
+            headers: { 'User-Agent': BitstampAuthenticatedClient.UserAgent }
+        });
+    };
+    
+    private requestFromEndpoint = <TResponse>(options: request.Options) => {
+        var defer = Q.defer<TResponse>();
+        request(options, (err, resp, body) => {
+            if (err) defer.reject(err);
+            else defer.resolve(JSON.parse(body));
+        });
+        return defer.promise;
+    };
+}
+
+class BitstampEntryGateway implements Interfaces.IOrderEntryGateway {
```

Comment:
Nothing here was tested. I was just basing off the API documentation https://www.bitstamp.net/api/